### PR TITLE
Generate "Ubershaders" that can handle all lighting configurations

### PR DESCRIPTION
**Pull request by hrydgard**
Instead of a ton of specific shaders for every little difference.

Applies to Vulkan, DX11 and GLES3.

This drastically reduces or removes the shader compile stutter that happens when a lot of new light setups are created, like on the first punch in Tekken 6.

There's more stuff that might benefit from being made dynamic like this. These branches are very cheap on modern GPUs since they're branching on a uniform variable, so no divergence.

I think we'll need to keep the old path too for gpus like Mali-450...

Tested in VK, D3D11, and OpenGL/ES.

~~Also! TODO: this need to somehow take doLight from shade mapping into account.~~

### CODE REVIEW

File: GPU/Common/ShaderUniforms.h [hunk 1]
```diff
@@ -95,6 +96,7 @@ R"(	vec4 u_ambient;
 	vec3 u_matdiffuse;
 	vec4 u_matspecular;
 	vec3 u_matemissive;
+    uint u_lightControl;  // light ubershader
```
Comment (unknownbrackets) [inline GPU/Common/ShaderUniforms.h:99]: nit: Indentation.

-[Unknown]

File: GPU/Common/ShaderId.cpp [hunk 2]
```diff
@@ -125,13 +128,17 @@ void ComputeVertexShaderID(VShaderID *id_out, u32 vertType, bool useHWTransform,
...
 			id.SetBits(VS_BIT_MATERIAL_UPDATE, 3, gstate.getMaterialUpdate());
 			id.SetBit(VS_BIT_LIGHTING_ENABLE);
-			// Light bits
-			for (int i = 0; i < 4; i++) {
-				bool chanEnabled = gstate.isLightChanEnabled(i) != 0;
-				id.SetBit(VS_BIT_LIGHT0_ENABLE + i, chanEnabled);
-				if (chanEnabled) {
-					id.SetBits(VS_BIT_LIGHT0_COMP + 4 * i, 2, gstate.getLightComputation(i));
-					id.SetBits(VS_BIT_LIGHT0_TYPE + 4 * i, 2, gstate.getLightType(i));
+			if (g_Config.bLightUberShader) {
```
Comment (unknownbrackets) [inline GPU/Common/ShaderId.cpp:131]: Should this also check bitwiseOps?

-[Unknown]

**Review by unknownbrackets — COMMENTED** [review]
Yeah, this is probably one of the better areas to "uber" combine.  It's a shame about Mali and older GPUs since this'll complicate things a bit... but seems worth it to me as well.

Wish we had lighting right before this, though...

-[Unknown]

**Comment by hrydgard** [comment]
Indeed would be nice to have lighting correct, but the duplicated changes will at least be very nearby and easy to port, once we find what's wrong.

And yeah, this will get a round of fixes before it's time to review, thanks for looking though :)

File: GPU/Common/ShaderId.cpp (hunk 2 above)
Comment (hrydgard) [reply GPU/Common/ShaderId.cpp:131]: Changed it so that the ubershader is now a GPU_FEATURE flag, and it's set depending on whether bitwise ops work.

**Comment by hrydgard** [comment]
GLES being a pedant about signed vs unsigned again, of course. Fixing.

**Comment by hrydgard** [comment]
Wait, forgot the shade mapping part.. it might be working but I need to check, heh.

**Comment by hrydgard** [comment]
Oh, the shade mapping (env mapping) is working, but there's some old cruft to clean out...

**Comment by hrydgard** [comment]
Fixed the test, should be mostly finished now, I think.

**Review by unknownbrackets — APPROVED** [review]

**Comment by Gamemulatorer** [comment]
This fixes https://github.com/hrydgard/ppsspp/issues/13480 and https://github.com/hrydgard/ppsspp/issues/15239 thank you!!! ;)

**Comment by hrydgard** [comment]
@Gamemulatorer #16111 plus #16112 might help a bit more.

**Pull request merged**